### PR TITLE
Deposit events history

### LINKED ISSUE #163: Eth1Notifier init from deposit contract
**Is your feature request related to a problem? Please describe.**
If Eth1Notifier is started after Deposit/Eth2Genesis events happens it will have wrong depositCount and chainStarted props and thus won't process events further.

**Describe the solution you'd like**
On start function, filter Deposit events from block at which contract is deployed, store them in db, store count in depositCount prop. Filter Eth2Genesis event, store in db and set chainStarted to true.

### PR DESCRIPTION
**Pull request by mpetrunic**
The logic is:
- if Eth2Genesis event occured
   - just subscribe to block updates (deposit events will be extracted in processBlock method I think)
- If there is no Eth2Genesis event
   - fetch all deposit events since contract deployed and store in database
   - subscribe to blocks, Deposit and Eth2Genesis events

Should resolve #163

### CODE REVIEW

File: src/eth1/impl/ethers.ts [hunk 1]
```diff
@@ -154,4 +157,39 @@ export class EthersEth1Notifier extends EventEmitter implements Eth1Notifier {
...
+    return logs.map((log) => {
+      const logDescription = this.contract.interface.parseLog(log);
+      return this.createDeposit(logDescription.values.dataHex, logDescription.values.indexHex);
+    });
+  }
+
+  private async getContractPastLogs(topics: string[]): Promise<Log[]> {
+    const filter = {
+      fromBlock: this.opts.depositContract.deployedAt,
+      toBlock: this.genesisBlockHash,
```
Comment (wemeetagain) [inline src/eth1/impl/ethers.ts]: I think these should be allowed to be changed, we may want to look at past logs from after deployedAt, and after genesisBlockHash

**Review by GregTheGreek — DISMISSED** [review]

File: src/eth1/impl/ethers.ts [hunk 2]
```diff
@@ -50,12 +51,23 @@ export class EthersEth1Notifier extends EventEmitter implements Eth1Notifier {
...
-      `Started listening on eth1 events on chain ${(await this.provider.getNetwork()).chainId}`
-    );
+    if(await this.isAfterEth2Genesis()) {
+      logger.info('Eth2Genesis event exits, started listening on eth1 block updates');
+      this.provider.on('block', this.processBlockHeadUpdate.bind(this));
+    } else {
+      const pastDeposits = await this.getPastDepositEvents();
+      await Promise.all(pastDeposits.map((pastDeposit) => {
+        return this.db.setGenesisDeposit(pastDeposit);
+      }));
```
Comment (wemeetagain) [inline src/eth1/impl/ethers.ts:63]: I think we we're missing deposits between genesis and now that we'll need to calculate proofs?

**Review by wemeetagain — DISMISSED** [review]

File: src/eth1/impl/ethers.ts (hunk 1 above)
Comment (mpetrunic) [reply src/eth1/impl/ethers.ts]: But we should obtain Deposit logs after Genesis event with syncing beacon blocks from other clients right?

File: src/eth1/impl/ethers.ts (hunk 2 above)
Comment (mpetrunic) [reply src/eth1/impl/ethers.ts:63]: This is case when genesis has not occured, we just have deposit logs till now. But now that you mentioned, ~~there is a bug toBlock in filter should default to latest I guess~~ (it defaults to latest if it's null^^)

Comment (GregTheGreek) [reply src/eth1/impl/ethers.ts:63]: I think @mpetrunic is correct, genesis hasn't fired

File: src/eth1/impl/ethers.ts (hunk 1 above)
Comment (mpetrunic) [reply src/eth1/impl/ethers.ts]: @wemeetagain Am I assuming this correctly? :)

Comment (wemeetagain) [reply src/eth1/impl/ethers.ts]: I think we still want to collect deposits from the eth1 chain directly as an additional data source for our validators. We'll be getting 'pending' (not in chain yet) deposits from the eth1 chain and maybe from other peers. What we have right now roughly is set up to do that (the event handler on L63), but if we don't sweep the deposit logs until "now", there's still the potential of missing pending deposits that are in eth1 blocks before "now", when the deposit event handler starts.

Comment (mpetrunic) [reply src/eth1/impl/ethers.ts]: Oh yeah, BeaconBlock doesn't contain DepositData array. I will expose method getPastDeposits(fromblock, toBlock?) right?

File: src/eth1/impl/ethers.ts [hunk 3]
```diff
@@ -125,6 +128,22 @@ export class EthersEth1Notifier extends EventEmitter implements Eth1Notifier {
     await this.db.deleteGenesisDeposits(genesisDeposits);
   }
 
+
+  public async getContractDeposits(
+    fromBlock: string | number  = this.opts.depositContract.deployedAt,
+    toBlock?: string | number
```
Comment (GregTheGreek) [inline src/eth1/impl/ethers.ts:133]: I would default this to `null` so you don't need to pass in null

File: src/eth1/impl/ethers.ts [hunk 4]
```diff
@@ -55,7 +55,10 @@ export class EthersEth1Notifier extends EventEmitter implements Eth1Notifier {
       logger.info('Eth2Genesis event exits, started listening on eth1 block updates');
       this.provider.on('block', this.processBlockHeadUpdate.bind(this));
     } else {
-      const pastDeposits = await this.getPastDepositEvents();
+      const pastDeposits = await this.getContractDeposits(
+        this.opts.depositContract.deployedAt,
+        null
```
Comment (GregTheGreek) [inline src/eth1/impl/ethers.ts]: remove this null

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: src/eth1/impl/ethers.ts (hunk 3 above)
Comment (mpetrunic) [reply src/eth1/impl/ethers.ts:133]: it is null by default. Notice the question mark

**Review by GregTheGreek — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**